### PR TITLE
add io.github.dweymouth.supersonic

### PR DESCRIPTION
**Pull request by anarcat**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### Comments

This is not quite as good as I would like:

- ~~it needs --network=share to build, because of the gigantic forest of Golang dependencies, which I haven't quite worked through yet~~ fixed
- it doesn't use Wayland, but so does the upstream binary release in my experience
- ~~the --device could be tighter~~ fixed
- it refers to the tip of the upstream git branch because the first beta has playback issues that were fixed there, breaking the "Stable releases, reproducible builds" requirement
- ~~there's a `.desktop` file inside the flatpak (`/app/share/applications/Supersonic.desktop`) but i'm not sure it's setup correctly~~ fixed
- ~~the app ID is `com.github.dweymouth.supersonic` because there's no github.io site for this project~~ fixed

This is my first flatpak ever and there might be some other snags in it that I'm not aware of, all feedback welcome!

A full history of the changes to this repo is also visible in:

https://github.com/anarcat/com.github.dweymouth.supersonic

### CODE REVIEW

**Comment by hfiguiere** [comment]
you need to create the manifest for the golang dependencies. See https://github.com/flatpak/flatpak-builder-tools/tree/master/go-get

For the desktop file. `rename-desktop-file` is probably what you are looking for

File: com.github.dweymouth.supersonic.yml [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+sdk: org.freedesktop.Sdk
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.golang
+command: supersonic
+
+finish-args:
+  # Wayland
+  - --socket=wayland
+  # from https://github.com/flathub/im.riot.Riot/blob/ff59f1ac426823e95444d67c763cd92523f1db4e/im.riot.Riot.yaml#L37
+  # Required for experimental wayland support
```
Comment (hfiguiere) [inline com.github.dweymouth.supersonic.yml:20]: if it is not out of the box it should be disabled.

File: com.github.dweymouth.supersonic.yml [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  # https://docs.flatpak.org/en/latest/electron.html for how it's done
+  # with electron apps. alternatively, mpv could be a "shared module",
+  # see:
+  # https://docs.flatpak.org/en/latest/manifests.html#shared-modules
+  # https://github.com/flathub/shared-modules
+  - name: mpv
+    buildsystem: meson
+    build-options:
+      env:
+        CFLAGS=-DNDEBUG=1
```
Comment (hfiguiere) [inline com.github.dweymouth.supersonic.yml:70]: use `cflags` instead of `env`

File: com.github.dweymouth.supersonic.yml [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+      # https://github.com/flatpak/flatpak-builder-tools/tree/master/go-get
+      # ... because it generates git links, which are much slower
+      #
+      # other examples:
+      # https://github.com/flathub/me.kozec.syncthingtk/blob/cf25336132b96c514f29d5a5322874e847aba150/syncthing.yaml
+      #
+      # also used:
+      # https://github.com/search?q=org%3Aflathub+++mpv&type=code
+      build-args:
+        - --share=network
```
Comment (hfiguiere) [inline com.github.dweymouth.supersonic.yml:187]: this is NOT allowed at all.

File: com.github.dweymouth.supersonic.yml [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib /app/bin/fyne package -os linux -name Supersonic -icon ./res/appicon-500.png
+      - tar -C /app --strip-components=1 -x -v -f /run/build/supersonic/Supersonic.tar.xz
+      - mv /app/share/applications/Supersonic.desktop /app/share/applications/com.github.dweymouth.supersonic.yml
+    sources:
+      - type: git
+        url: https://github.com/dweymouth/supersonic/
+        commit: 440dee15ac1bb42fcd899578576098542ffc4a19
+    post-install:
+      # workaround for go annoyance
+      - chmod -R u+w $FLATPAK_BUILDER_BUILDDIR
```
Comment (hfiguiere) [inline com.github.dweymouth.supersonic.yml:212]: I don't know why you need this.

File: com.github.dweymouth.supersonic.yml [hunk 5]
```diff
@@ -202,7 +202,7 @@ modules:
       - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib go build -v .
       - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib /app/bin/fyne package -os linux -name Supersonic -icon ./res/appicon-500.png
       - tar -C /app --strip-components=1 -x -v -f /run/build/supersonic/Supersonic.tar.xz
-      - ls -al /app/bin/supersonic
+      - mv /app/share/applications/Supersonic.desktop /app/share/applications/com.github.dweymouth.supersonic.yml
```
Comment (hfiguiere) [inline com.github.dweymouth.supersonic.yml:205]: this is what `rename-desktop-file` is for.

File: com.github.dweymouth.supersonic.yml (hunk 4 above)
Comment (anarcat) [reply com.github.dweymouth.supersonic.yml:212]: this is to work around permission problems with the golang environment. at first i just couldn't build the flatpak at all. it would fail near the very end with:

    Error: module supersonic: unlinkat: Permission denied

I found this hack on https://www.reddit.com/r/flatpak/comments/n8y3he/comment/gxla9og/?context=3

One of the build log failures looked like:

https://paste.anarc.at/publish/2023-04-13-WKiMhZ8aicJYxQkrjVQTgaioagwkf5EjSRY42P0TkZw/clipboard.txt

and back then the YAML file looked like:

https://paste.anarc.at/publish/2023-04-13-WKiMhZ8aicJYxQkrjVQTgaioagwkf5EjSRY42P0TkZw/com.github.dweymouth.supersonic.yml

I've had to squash so many of those things to even get the app to compile that I haven't dared to remove this yet, but maybe it's worth a try!

**Comment by anarcat** [comment]
> you need to create the manifest for the golang dependencies. See https://github.com/flatpak/flatpak-builder-tools/tree/master/go-get

So I looked at that tool, and couldn't get it to work. It says to pass the `build-dir` but then crashes with a Python backtrace because it can't find the `src` directory.

After much wrangling, I managed to make it somewhat work, but it only downloaded parts of the manifests.

If you are familiar with golang, you'll notice the build is not quite straightfoward. It's in multiple stages and has a *lot* of dependencies. All of Fyne, in particular, is a huge lot, and the author uses a fork of it so I am not sure I can manage to just feed a list of sources in that way.

Also, when I tried to bundle at least a *part* of the deps, I couldn't place them in the right path for golang to find them...

I would very much need help with this part. I've been working on this for almost a full day now and I'm just about to give up. I've managed to make it build and run and I can play audio, so that's good enough for me for now. It would be too bad for this to die at the last gate because of administrative limitations like this...

In other words, I would love if someone more familiar with golang + flatpak could sweep in and cleanup that mess here. It doesn't make it any easier that there's no standard and reliable way to do this... all projects seem to have their own way to do this, and none that I have found so far actually use the go-get hack...

File: com.github.dweymouth.supersonic.yml [hunk 6]
```diff
@@ -0,0 +1,212 @@
+# example flatpaks:
+# https://github.com/search?q=org%3Aflathub+golang+&type=code&p=2
+# https://github.com/search?q=org%3Aflathub+GLFW+&type=code
+#
+# followup on:
+# https://discourse.flathub.org/t/supersonic-lightweight-cross-platform-desktop-client-for-subsonic-music-servers/3984/2
+# https://github.com/dweymouth/supersonic/issues/86
+app-id: com.github.dweymouth.supersonic
```
[the remark this answers is not in the document]
Comment (anarcat) [reply com.github.dweymouth.supersonic.yml]: i was under the impression that the point of the app-id was that it was a space the upstream controlled. since there was no github.io site, i thought it was better to use github.com since that is where the site actually is, but i'll trust you on this and i changed it to io.github. :)

File: io.github.dweymouth.supersonic.yml [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+  # https://docs.flatpak.org/en/latest/electron.html for how it's done
+  # with electron apps. alternatively, mpv could be a "shared module",
+  # see:
+  # https://docs.flatpak.org/en/latest/manifests.html#shared-modules
+  # https://github.com/flathub/shared-modules
+  - name: mpv
+    buildsystem: meson
+    build-options:
+      cflags: -DNDEBUG=1
+      env:
```
Comment (hfiguiere) [inline io.github.dweymouth.supersonic.yml:79]:
```suggestion
```

File: io.github.dweymouth.supersonic.yml [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+          sha256: f0da0bbfba476c16ae3e1cfd862256d30915911f7abaa1b16ce62ee653192784
+          x-checker-data:
+            type: anitya
+            project-id: 1560
+            stable-only: true
+            url-template: https://github.com/libass/libass/releases/download/$version/libass-$version.tar.xz
+        - type: script
+          commands:
+            - autoreconf -fiv
+          dest-filename: autogen.sh
```
Comment (hfiguiere) [inline io.github.dweymouth.supersonic.yml:123]:
```suggestion
        - type: shell
          commands:
            - autoreconf -fiv
```

File: io.github.dweymouth.supersonic.yml [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+          mirror-urls:
+            - https://git.ffmpeg.org/ffmpeg.git
+          commit: ea3d24bbe3c58b171e55fe2151fc7ffaca3ab3d2
+          tag: n6.0
+          x-checker-data:
+            type: git
+            tag-pattern: ^n([\d.]{3,7})$
+
+  # this is used to bundle the final supersonic command correctly and
+  # provides the /app/bin/fyne command to the next module
```
Comment (hfiguiere) [inline io.github.dweymouth.supersonic.yml:172]: so you probably want a cleanup rule `*` to remove it afterwards

File: io.github.dweymouth.supersonic.yml (hunk 7 above)
Comment (anarcat) [reply io.github.dweymouth.supersonic.yml]: fixed thanks.

File: io.github.dweymouth.supersonic.yml (hunk 9 above)
Comment (anarcat) [reply io.github.dweymouth.supersonic.yml:171]: cleanup rules run at the very end of all the modules? i imagined they ran at the end of the module build...

File: io.github.dweymouth.supersonic.yml [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+        - --share=network
+    build-commands:
+      - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib go build -v .
+      - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib /app/bin/fyne package -os linux -name Supersonic -icon ./res/appicon-500.png
+      - tar -C /app --strip-components=1 -x -v -f /run/build/supersonic/Supersonic.tar.xz
+    sources:
+      - type: git
+        url: https://github.com/dweymouth/supersonic/
+        commit: 440dee15ac1bb42fcd899578576098542ffc4a19
+    rename-desktop-file: /app/share/applications/Supersonic.desktop
```
Comment (hfiguiere) [inline io.github.dweymouth.supersonic.yml:217]: that doesn't belong here. See `man flatpak-manifest`

Also the syntax is wrong.

File: com.github.dweymouth.supersonic.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (hfiguiere) [reply com.github.dweymouth.supersonic.yml]: actually `io.github.USER_OR_ORG` is always granted. Also this allow web based verification schemes.

**Comment by hfiguiere** [comment]
and please stop mentioning me in commits. I don't need more github spam

**Comment by anarcat** [comment]
thanks for all the feedback @hfiguiere (and hi, btw, long time no see!)... i've implemented most of your suggestions, i think! the only remaining problem (and it's a huge one) is trying to find a way to automatically list all the golang dependencies (no way anyone is going to run through this by hand) and add a sources for that...

as I said, last time i tried go-get for that, it failed, would love to be proven wrong on that. :)

**Comment by anarcat** [comment]
> and please stop mentioning me in commits. I don't need more github spam

ouch, sorry.

File: com.github.dweymouth.supersonic.yml (hunk 3 above)
Comment (anarcat) [reply com.github.dweymouth.supersonic.yml]: fixed!

File: com.github.dweymouth.supersonic.yml [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+  - name: supersonic
+    buildsystem: simple
+    build-options:
+      append-path: /usr/lib/sdk/golang/bin
+      env:
+        GOBIN: /app/bin
+      # TODO: remove this and get each dependency properly, then add
+      # `GO111MODULE: off` above
+      build-args:
+        - --share=network
```
[the remark this answers is not in the document]
Comment (anarcat) [reply com.github.dweymouth.supersonic.yml]: fixed!

**Comment by anarcat** [comment]
after a long struggle, i have figured out a way to pull the sources list out automatically. it has not been easy, but i think it works. thankfully, fyne itself has a vendored tarball so i could just use that directly, without any other deps. 

this is now ready for final review and merge, thanks for the initial review again.

**Comment by hfiguiere** [comment]
bot, build io.github.dweymouth.supersonic

**Comment by hfiguiere** [comment]
the AppStream file can't be found.

**Comment by anarcat** [comment]
yep, i see that. didn't realize those were mandatory. i'm kind of exhausted now, so i'll ask upstream if they can kindly provide one.

**Comment by hfiguiere** [comment]
```
WARNING: Icon referenced in desktop file but not exported: Supersonic
```
this icon name doesnt' match.

File: io.github.dweymouth.supersonic.yml [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+        - --enable-libmp3lame
+        - --enable-libopus
+        - --enable-libtheora
+        - --enable-libv4l2
+        - --enable-libvorbis
+        - --enable-libvpx
+      sources:
+        - type: git
+          url: https://github.com/FFmpeg/FFmpeg.git
+          mirror-urls:
```
Comment (hfiguiere) [inline io.github.dweymouth.supersonic.yml:163]: `(flatpak-builder:2): flatpak-builder-WARNING **: 20:11:25.885: Unknown property mirror-urls for type BuilderSourceGit`

Comment (anarcat) [reply io.github.dweymouth.supersonic.yml:163]: odd. that's from: 

https://github.com/flathub/com.icanblink.blink/blob/75a76c08017c41833a30dc879f00c1aec9662a28/modules/io.mpv.Mpv.json#L57

**Comment by hfiguiere** [comment]
> ```
> WARNING: Icon referenced in desktop file but not exported: Supersonic
> ```
> 
> this icon name doesnt' match.

`usr/share/pixmaps/Supersonic.png` is the wrong path and the wrong name.

File: io.github.dweymouth.supersonic.yml (hunk 12 above)
Comment (hfiguiere) [reply io.github.dweymouth.supersonic.yml]: they don't read the logs and we don't make these fatal errors.

File: io.github.dweymouth.supersonic.yml [hunk 13]
```diff
@@ -13,6 +13,7 @@ sdk-extensions:
   - org.freedesktop.Sdk.Extension.golang
 command: supersonic
 rename-desktop-file: Supersonic.desktop
+rename-icon: Supersonic.png
```
Comment (hfiguiere) [inline io.github.dweymouth.supersonic.yml:16]:
```suggestion
rename-icon: Supersonic
``` 

no extension. That will also fix the .desktop

but the icon is still in the wrong path.

Comment (anarcat) [reply io.github.dweymouth.supersonic.yml:16]: > but the icon is still in the wrong path.

what would be the proper path? here it's in `/appshare/pixmaps/Supersonic.png`, what's the correct one?

Comment (hfiguiere) [reply io.github.dweymouth.supersonic.yml]: it has to follow the icon theme spec https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html#directory_layout

minimum size is 128x128.

Comment (hfiguiere) [reply io.github.dweymouth.supersonic.yml]: ideally it would also be fixed upstream.

Comment (anarcat) [reply io.github.dweymouth.supersonic.yml]: i tried to move the icon to the right location in [0ea2517](https://github.com/flathub/flathub/pull/4073/commits/0ea2517ce7d1fe6ae6734afa2fca96511548032c) but now the build fails with

```
Error: icon Supersonic not found below /home/anarcat/src/flathub/.flatpak-builder/rofiles/rofiles-qg3xRa/files/share/icons
```

not sure what i am doing wrong now...

File: io.github.dweymouth.supersonic.yml [hunk 14]
```diff
@@ -197,6 +197,8 @@ modules:
       - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib go build -v .
       - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib /app/bin/fyne package -os linux -name Supersonic -icon ./res/appicon-500.png
       - tar -C /app --strip-components=1 -x -v -f /run/build/supersonic/Supersonic.tar.xz
+      - mkdir -p /app/share/icons
+      - mv /app/share/pixmaps/Supersonic.png /app/share/icons
```
Comment (hfiguiere) [inline io.github.dweymouth.supersonic.yml:201]: it should be `/app/share/icons/hicolor/SIZExSIZE/apps`

with `SIZE` being the size of the icon.

Comment (anarcat) [reply io.github.dweymouth.supersonic.yml:201]: i tried this:

```
modified   io.github.dweymouth.supersonic.yml
@@ -197,8 +197,8 @@ modules:
       - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib go build -v .
       - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib /app/bin/fyne package -os linux -name Supersonic -icon ./res/appicon-500.png
       - tar -C /app --strip-components=1 -x -v -f /run/build/supersonic/Supersonic.tar.xz
-      - mkdir -p /app/share/icons
-      - mv /app/share/pixmaps/Supersonic.png /app/share/icons
+      - mkdir -p /app/share/icons/hicolor/520x520/
+      - mv /app/share/pixmaps/Supersonic.png /app/share/icons/hicolor/520x520/
     sources:
       - type: git
         url: https://github.com/dweymouth/supersonic/
```

and still got this error: 

```
Error: icon Supersonic not found below /home/anarcat/src/flathub/.flatpak-builder/rofiles/rofiles-qg3xRa/files/share/icons
```

is there a standard set of sizes the icon should live in?

Comment (hfiguiere) [reply io.github.dweymouth.supersonic.yml:201]: 520x520 is not a standard size. also it's 512 maximum.

Comment (anarcat) [reply io.github.dweymouth.supersonic.yml:201]: i filed https://github.com/dweymouth/supersonic/issues/132 upstream. i have tried to throw the 520x520 icon in the 512x512 directory and that still doesn'T work... i am beginning to suspect the rename call is what is broken here, unless it's actually checking the file resolution?

Comment (hfiguiere) [reply io.github.dweymouth.supersonic.yml:201]: also it's still the wrong path. missing `apps` as the last directory.

Comment (anarcat) [reply io.github.dweymouth.supersonic.yml:201]: it is, and i fixed this.

**Comment by anarcat** [comment]
> the AppStream file can't be found.

this is now fixed too.

File: io.github.dweymouth.supersonic.yml (hunk 13 above)
Comment (anarcat) [reply io.github.dweymouth.supersonic.yml]: fixed and reported upstream as https://github.com/dweymouth/supersonic/issues/132

File: io.github.dweymouth.supersonic.yml [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+      env:
+        GOBIN: /app/bin
+        GO11MODULE: off
+    build-commands:
+      - mkdir -p /app/share/icons/hicolor/512x512/apps/
+      - mv $FLATPAK_ID.png /app/share/icons/hicolor/512x512/apps/
+      - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib go build -v .
+      - . /usr/lib/sdk/golang/enable.sh && CGO_LDFLAGS=-L/app/lib /app/bin/fyne package -os linux -name Supersonic -icon ./res/appicon-500.png
+      - tar -C /app --strip-components=1 -x -v -f /run/build/supersonic/Supersonic.tar.xz
+      - sed -i s/Icon=Supersonic/Icon=$FLATPAK_ID/ /app/share/applications/Supersonic.desktop
```
Comment (hfiguiere) [inline io.github.dweymouth.supersonic.yml:201]: please use `desktop-file-edit` instead of sed.

File: io.github.dweymouth.supersonic.yml [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+      # io.mpv.Mpv has a neat trick to automatically update the
+      # version too, see:
+      # https://github.com/flathub/io.mpv.Mpv/blob/master/io.mpv.Mpv.yml#L606-L617
+      - mkdir -p /app/share/appdata
+      - mv $FLATPAK_ID.appdata.xml /app/share/appdata/$FLATPAK_ID.appdata.xml
```
Comment (hfiguiere) [inline io.github.dweymouth.supersonic.yml:206]:
```suggestion
      - install -Dm644 $FLATPAK_ID.appdata.xml /app/share/appdata/$FLATPAK_ID.appdata.xml
```

**Comment by hfiguiere** [comment]
bot, build io.github.dweymouth.supersonic

**Comment by hfiguiere** [comment]
AppStream file fail to validate:
```
• tag-missing           : <content_rating> required [use https://odrs.gnome.org/oars]
• style-invalid         : Not enough <screenshot> tags, minimum is 1
```

**Comment by hfiguiere** [comment]
bot, build io.github.dweymouth.supersonic

**Comment by anarcat** [comment]
> AppStream file fail to validate:
> 
> ```
> • tag-missing           : <content_rating> required [use https://odrs.gnome.org/oars]
> • style-invalid         : Not enough <screenshot> tags, minimum is 1
> ```

fixed.

i think i whacked all the moles, again... so:

bot, build io.github.dweymouth.supersonic

i bet that doesn't work for me does it? :)

is there a way for me to run those linting tests? i understand i can run validate on the appstream data now, but there's so many of those little things, and that's just one of them... it would be nice to have builder or some other tool that would run All The Checks at once.. e.g. i see now that flatpak-builder emits some warnings, but those can get lost in the long output...

anyway, should be cleared for another build run i guess.

**Comment by anarcat** [comment]
> To test this build, install it from the testing repository:

tested the build, it works.

**Comment by anarcat** [comment]
is there anything else i need to do here? :)
